extensions/wikia/JSMessages/JSMessages.class.php

```diff
@@ -124,191 +81,142 @@ static private function resolveMessagesPattern($pattern) {
+				$messageNames = $loader->getModule( self::RL_MODULE_PREFIX . $packageName )->getMessages();
+				foreach ( $messageNames as $msg ) {
+					$messages[$msg] = wfMessage( $msg )->inLanguage( $wg->Lang )->plain();
 				}
 			}
 		}
 
-		wfProfileOut(__METHOD__);
-		return $ret;
+		return $messages;
 	}
 
 	/**
-	 * Get messages from given packages
-	 *
-	 * Packages with widlcard matching can only be queued in EXTERNAL mode. Doing so in INLINE mode will cause
-	 * PHP fatal error (BugId:18482)
+	 * Given a message name ending with a wildcard (*), get a list of all matching message names
 	 *
-	 * @param array $packages - list packages names
-	 * @param boolean $allowWildcards - can packages with wildcard be added?
-	 * @return array - key/value array of messages
+	 * @param string $pattern message name pattern ending with a *
+	 * @return array list of matching message names
 	 */
-	static public function getPackages($packages, $allowWildcards = true) {
-		$messages = array();
-
-		foreach($packages as $packageName) {
-			$packageMessages = self::getPackage($packageName, $allowWildcards);
-
-			if (is_array($packageMessages)) {
-				$messages = array_merge($messages, $packageMessages);
+	static private function getMatchingMessagesForPattern( $pattern ) {
+		$pattern = substr( $pattern, 0, -1 );
+		$patternLen = strlen( $pattern );
+
+		$messages = self::getAllMessageKeys( Language::factory( F::app()->wg->Lang->getCode() ) );
+		$matching = [];
+		foreach( $messages as $msg ) {
+			if (substr( $msg, 0, $patternLen ) === $pattern) {
+				$matching[] = $msg;
 			}
 		}
 
-		return $messages;
+		return $matching;
 	}
 
 	/**
-	 * Emit messages from the queue as:
-	 *   - JS object in <head> section of the page (INLINE mode)
-	 *   - JS requested via <script> tag at the bottom of the page (EXTERNAL mode)
+	 * Expand wildcards in the message list of a given package.
+	 *
+	 * @param array $keys list of message keys, including wildcards
+	 * @return array list of message keys with all wildcards expanded
 	 */
-	static public function onWikiaSkinTopScripts( &$vars, &$scripts, $skin) {
-		wfProfileIn(__METHOD__);
-		self::log(__METHOD__, 'preparing list of inline messages...');
-
-		// get items to be rendered as a variable in <head> section
-		$packages = self::$queue['inline'];
-
-		if (!empty($packages)) {
-			$vars['wgMessages'] = self::getPackages($packages, false /* don't allow wildcards in INLINE mode (BugId:18482) */);
-		}
-
-		self::log(__METHOD__, 'preparing list of external packages...');
-
-		$url = self::getExternalPackagesUrl();
-
-		if ($url != "") {
-			// request a script
-			F::app()->wg->Out->addScript(Html::linkedScript($url));
+	static private function expandWildcards( $keys ) {
+		$list = [];
+		if ( is_array( $keys ) ) {
+			// Expand any wildcards, otherwise just push the message onto the list
+			foreach ( $keys as $message ) {
+				if ( substr( $message, -1 ) == '*' ) {
```

Comment:
If we move this conditional into `getMatchingMessagesForPatter`, and make it return $message when * is not find it code will look nicer (less cyclometric)
In this case, method names should be changed for more proper one.